# 아이템 9. try-finally보다는 try-with-resources를 사용하라

자바 라이브러리에는 close를 호출해 직접 닫아줘야 하는 자원이 많다. 이런 자원을 다룰 때는 try-finally 블록을 사용해야 한다. 하지만 try-finally 블록은 너무 지저분하고, 실수하기 쉽다. 그래서 자바 7부터는 try-with-resources라는 기능을 제공한다. 이 기능을 사용하면 try-finally 블록을 더 간결하고, 올바르게 사용할 수 있다.

```java
// 너무 더럽다
static void copy(String src, String dst) throws IOException {
    InputStream in = new FileInputStream(src);

    try {
        OutputStream out = new FileOutputStream(dst);
        try {
            byte[] buf = new byte[BUFFER_SIZE];
            int n;
            while ((n = in.read(buf)) >= 0)
                out.write(buf, 0, n);
        } finally {
            out.close();
        }
    } finally {
        in.close();
    }
}
```

위 예제는 try-finally를 잘 썼음에도 미묘한 결점이 있다. 바로 `out.close`에서 예외가 발생하면, `in.close`가 호출되지 않는다는 것이다.

이런 문제를 해결하기 위해 자바 7부터는 try-with-resources라는 기능을 제공한다. 이 구조를 사용하려면 해당 자원이 `AutoCloseable` 인터페이스를 구현해야 한다. `AutoCloseable` 인터페이스는 단 하나의 메서드를 정의하고 있다.

보통의 try-catch문 처럼 try-with-resources도 catch 블록을 가질 수 있다. 이 catch 블록은 try 블록에서 발생한 예외와 try-with-resources 블록에서 발생한 예외를 모두 처리한다.
